File: typescript/binary-search/binary-search.ts

export function find(haystack: number[], needle: number): number {
  let min = 0
  let max = haystack.length - 1
  let i: number

  while (min <= max) {
    i = Math.floor((max + min) / 2)

    if (haystack[i] === needle) return i
      if (haystack[i] > needle) {
        max = i - 1
      } else {
        min = i + 1
      }
  }

  throw new Error('Value not in array')
}

File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find (input: int[]) (value: int) =
    let rec search (min: int) (max: int) =
        if min > max then None
        else
            let mid = (max + min) / 2
            match input[mid] with
            | v when v = value -> Some mid
            | v when v < value -> search (mid + 1) max 
            | v when v > value -> search min (mid - 1)
            | _ -> failwith "invalid"

    search 0 (input.Length - 1)
